### ISSUE TITLE
CollectionView Drag & Drop Item Reordering Spec

### ISSUE THREAD

**Issue by billvenhaus** [dotnet/maui#3767]
# Description

Support drag-and-drop item reordering in the `CollectionView`.

The `CollectionView` handlers should use the built-in item reordering functionality provided by the platforms:
- On Android, attach the [ItemTouchHelper](https://developer.android.com/reference/androidx/recyclerview/widget/ItemTouchHelper) to the RecyclerView.
- On iOS, follow the procedure as described in [Reordering Items Interactively](https://developer.apple.com/documentation/uikit/uicollectionview#overview). Use a gesture recognizer make calls to [beginInteractiveMovementForItem](https://developer.apple.com/documentation/uikit/uicollectionview/1618019-begininteractivemovementforitem), [updateInteractiveMovementTargetPosition](https://developer.apple.com/documentation/uikit/uicollectionview/1618079-updateinteractivemovementtargetp), & [endInteractiveMovement](https://developer.apple.com/documentation/uikit/uicollectionview/1618082-endinteractivemovement)
- On Windows, enable reordering by setting [AllowDrop](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.uielement.allowdrop?view=winui-3.0) & [CanReorderItems](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.controls.listviewbase.canreorderitems?view=winui-3.0) to `true`.

Reordering should require the `CollectionView.ItemsSource` be bound to an `IList`. The platform handlers will use the `Remove` & `Insert` methods on the `IList` to complete the reorder operation.
 
Example
https://github.com/billvenhaus/ReorderableCollectionView.Maui

![reorderablecollectionview_ios](https://user-images.githubusercontent.com/6692379/146122143-58101310-452a-4045-aac4-9ddba2af9a45.gif)

# (Public) API Changes

Place the new reorder functionality in a separate class that the `CollectionView` can inherit. 

## CollectionView

```c#
public class CollectionView : ReorderableItemsView
{
}
```

## ReorderableItemsView 

```c#
public class ReorderableItemsView : GroupableItemsView
{
	public static readonly BindableProperty CanMixGroupsProperty;
	public bool CanMixGroups { get; set; }

	public static readonly BindableProperty CanReorderItemsProperty;
	public bool CanReorderItems { get; set; }

	public event EventHandler ReorderCompleted;
}
```

## Properties

| API | Description |
| ------------- | ------------- |
| `CanReorderItems`| Gets or sets a value that indicates whether items in the view can be reordered through user interaction. |
| `CanMixGroups`| Gets or sets a value that indicates whether items from different groups can be mixed during reordering. |

## Events

| API | Description |
| ------------- | ------------- |
| `ReorderCompleted`| Raised whenever reordering user interaction completes. |

# Usage Scenarios

### Basic Item Reordering
The developer has an app which manages the playlist for a media player. The user needs to be able to make changes to the track order by rearranging the songs with drag-and-drop. How can the developer achieve this UI with the proposed API?

In order to achieve this behavior, the developer should set the `CollectionView.CanReorderItems` property to `true`. They can be notified of changes to the song order by subscribing to the `ReorderCompleted` event. At that point the developer can save the changes to the playlist storage location. 

### Reordering with Grouped Sources (Mixing Items)
The developer has a grouped data source & wishes to mix items. Items in group A can be placed in group B, and items in group B can be placed in group A. How can the developer achieve this with the proposed API?

In order to achieve this behavior, the developer should set both `CollectionView.CanReorderItems` & `CollectionView.CanMixGroups` to `true`.

### Detailed Events
The `ReorderCompleted` is a basic `EventHandler` that doesn't provide any information about the item that moved. How can the developer determine which item moved & where it moved to?

The developer should bind the `CollectionView.ItemsSource` to an `ObservableCollection` if they need to know specific details about the item that moved. The `CollectionChanged` event of the `ObservableCollection` can provide that information.

# Backward Compatibility

Minimum API Levels:
- Available on all API levels.

Breaking Changes:
- No breaking changes.

Unsupported Platforms:
- Windows reordering only works with ObservableCollections
- Windows reordering does not support grouped sources (see [here](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.controls.listviewbase.canreorderitems?view=winui-3.0))

# Difficulty

Medium

**Comment by AmrAlSayed0** [comment]
> Reordering should require the `CollectionView.ItemsSource` be bound to an IList.
Weird, one would think that `ObservableCollection` is the obvious choice here. Any reason for not having it be a `INCC` instead?

**Comment by billvenhaus** [comment]
Thanks!  Yes, it supports `ObservableCollection` too. Those actually complicated this task a little bit on Android & iOS. I had to add code to temporarily suspend notifications from `INCC` to the platform's `IItemsViewSource` while a move is in progress. Otherwise it disrupts the reorder operation. The `IItemsViewSource` is set up to refresh the UI whenever the backing`INCC` changes. Well in this case the UI is already aware of the move because it orchestrated the move.

Windows however does require an `ObservableCollection`. For that reason, I did seriously consider making `ObservableCollection` as a requirement in the spec. So perhaps the spec should be changed? The main reason I chose `IList` in the spec is that it offers broader support on Android & iOS. Forcing those platforms to use an `ObservableCollection` when it really isn't needed could result in some unnecessary overhead.
